### ISSUE TITLE
Web3.js transaction serialize => deserialize => serialize can produce "Signature verification failed" due to bug in Message serialization

### ISSUE THREAD

**Issue by ChewingGlass** [solana-labs/solana#21722]
#### Problem

The logic here https://github.com/solana-labs/solana-web3.js/blob/74007a55afa29b36e0535bb9bcff3b64dda9b19e/src/transaction.ts#L261-L281

Can lead to different orderings to accountMetas after serializing and deserializing.  The "content" of the transaction is the same, however the accountKeys flipped around. So long as the indices still hold true for writable accounts, the transaction will function the same.

The issue is that this is then serialized and used to create a transaction signature. So this non-determinism can mean that a previously correct signature is now incorrect. 

Code to reproduce:

```js
    const signer = Keypair.generate();
    const acc0Writable = Keypair.generate();
    const acc1Writable = Keypair.generate();
    const acc2Writable = Keypair.generate();
    const t0 = new Transaction({
      recentBlockhash: "HZaTsZuhN1aaz9WuuimCFMyH7wJ5xiyMUHFCnZSMyguH",
      feePayer: signer.publicKey
    });
    t0.add(new TransactionInstruction({
      keys: [{
        pubkey: signer.publicKey,
        isWritable: true,
        isSigner: true
      }, {
        pubkey: acc0Writable.publicKey,
        isWritable: true,
        isSigner: false
      }],
      programId: Keypair.generate().publicKey
    }))
    t0.add(new TransactionInstruction({
      keys: [{
        pubkey: acc1Writable.publicKey,
        isWritable: false,
        isSigner: false
      }],
      programId: Keypair.generate().publicKey
    }))
    t0.add(new TransactionInstruction({
      keys: [{
        pubkey: acc2Writable.publicKey,
        isWritable: true,
        isSigner: false
      }],
      programId: Keypair.generate().publicKey
    }))
    t0.add(new TransactionInstruction({
      keys: [{
        pubkey: signer.publicKey,
        isWritable: true,
        isSigner: true
      },{
        pubkey: acc0Writable.publicKey,
        isWritable: false,
        isSigner: false
      },  {
        pubkey: acc2Writable.publicKey,
        isWritable: false,
        isSigner: false
      }, {
        pubkey: acc1Writable.publicKey,
        isWritable: true,
        isSigner: false
      }],
      programId: Keypair.generate().publicKey
    }))
    t0.partialSign(signer);
    const t1 = Transaction.from(t0.serialize());
    t1.serialize()
```

A workaround is to, before pre-signing the transaction, run Transaction.from(tx.serialize({ requireAllSignatures: false })). Then sign the transaction and serialize. This ensures the ordering doesn't get jumbled. 

#### Proposed Solution

Add a default sorting here based on the alphabetical (or just numerical) sorting of the public keys so that this function is deterministic.

https://github.com/solana-labs/solana-web3.js/blob/74007a55afa29b36e0535bb9bcff3b64dda9b19e/src/transaction.ts#L261

**Comment by ChewingGlass** [comment]
Created a PR to fix this issue

**Comment by ponWeb** [comment]
Hey, do you think you can help me? https://github.com/solana-labs/solana/issues/22391
I think you have described my problem. But Im using offline signing, so I cant user partialSign because I dont have signer object. If you will refactor my code to needed or just give advice I would be really happy!

**Comment by z3000z3000z3000** [comment]
Hmm. After this change, I can't sign offline transaction, generating by solana cli tool.

As soon as solana cli tool generate transaction with this order:
```
                const transaction2 = new web3.Transaction({
                    recentBlockhash: nonceBlockHash, //recentBlockhash.blockhash,
                    nonceInfo: {
                        nonce: nonceBlockHash,
                        nonceInstruction: new web3.TransactionInstruction({
                            keys: [
                                {isSigner: false, isWritable: true, pubkey: new web3.PublicKey(noncePubkey)},
                                {isSigner: false, isWritable: false, pubkey: new web3.PublicKey("SysvarRecentB1ockHashes11111111111111111111")},
                                {isSigner: true, isWritable: true, pubkey: serverPK}
                            ],
                            programId: new web3.PublicKey("11111111111111111111111111111111"),
                            data: toBuffer([4,0,0,0])
                        })
                    },
                    feePayer: mypk
                }).add(
                    web3.SystemProgram.transfer({
                        fromPubkey: wallet.publicKey,
                        toPubkey: serverPK,
                        lamports: price,
                    })
                );
```
this is example of transaction(done by this code with version 1.31): 
https://explorer.solana.com/tx/5y5tbJVThh6p4KcgvaF6jF2KcctF4pN2Q7jFdSSd87T81k4MxGVku6mXywtRTC1NxULCmiFfjE4sHCAa4QQeV7uP?cluster=devnet

**Comment by z3000z3000z3000** [comment]
Created PR. I think it could work. But I don't sure, I can't build it locally. If I got problem correct - problem with unstable sort in js.

**Comment by antoineherzog** [comment]
Hi @jstarry!
Is the order of serialization part of the Solana spec officially?
We have some problem here when we need to communicate between Dart Library and Js Library:
https://github.com/cryptoplease/cryptoplease-dart/pull/184

**Comment by jstarry** [comment]
I don't think it's useful to enforce or follow any kind of canonical account key sorting method. The way `Transaction.populate` is decompiling the message account keys is losing the ordering used by the processed transaction so I think we have to get rid of that.

**Comment by ChikinDeveloper** [comment]
Hi! 
Totally agree that it isn't useful to enforce it, but don't you think there should be some kind of deterministic developer standard for transaction serializations, in order to prevent this kind of issue? 
Right now there's no guarantee that this implementation won't change in the future.

**Comment by antoineherzog** [comment]
What would be the next step?
There are incompatibilities between librairies (Js, Dart etc..) when we are passing a serialization between different librairies.

**Comment by t-nelson** [comment]
IMO the next step is what jstarry mentioned above, deser needs to not be reordering anything.  That's the fix.

Generally I question the motivation for doing a ser -> deser -> ser loop.  There's not really any valid reason to do that.  Only a compiled `Message` should be passed around for signing. Perhaps `partialSign` is implying it's intended for things that it's not

**Comment by z3000z3000z3000** [comment]
> 
Thanks. then, e.g. what a canonical way to sign offline nonced transaction, like here from web3?
https://docs.solana.com/offline-signing/durable-nonce

With cli I just write (in code with rust lib the same)
```
servsign3=`solana transfer $dest 0.1 -u devnet  --blockhash $nonceblockhash --nonce $nonceaddr    --from $clientaddr --fee-payer $clientaddr  --nonce-authority $server --sign-only | grep $serveraddr`

clientsign1=`$solana transfer $dest 0.1 -u devnet  --blockhash $nonceblockhash --nonce $nonceaddr    --from $clientaddr --nonce-authority $serveraddr -k $client --sign-only | grep $clientaddr`

solana transfer $dest 0.1 -u devnet  --blockhash $nonceblockhash --nonce $nonceaddr --from $clientaddr --fee-payer $clientaddr --allow-unfunded-recipient --nonce-authority $serveraddr --signer $servsign3 --signer $clientsign1
```

But, if one ot this signature are generating by web3 - I got different order of accounts - diffrent transaction and Signature verification failed.

**Comment by z3000z3000z3000** [comment]
also, this reordering broke order of transaction for complex transaction. E.g.
for this transaction https://explorer.solana.com/tx/3mr8xU5ivLNTNv5MwkLWhvVvAKg69bEqNJgaJWwe1tmcivgnVtFBkbn1DGVjSPQUnw94oD66xEbi3ofQBGxQh5S1

we always have first account - is payer, second is new NFT address &etc.
But with this fix - nft addr could be on 3 or 4 place, for example. So, order depends not only the transaction, but also of the base58 string of accounts.

**Comment by Sotatek-DucPham** [comment]
I faced the same problem with the latest version of @solana/web3.js. So I have a temporary solution

Instead of Sign => Serialize => Deserialize => Serialize. I choose Serialize => Deserialize => Sign => Serialize.

**Comment by jordansexton** [comment]
I dug into the code and came up with what I think is causing this. Related: #23641, #23610, https://github.com/solana-labs/wallet-adapter/issues/350

1. https://github.com/solana-labs/solana/blob/17cc095d2892654225b5efe91b55c19eb932bfd3/web3.js/src/message.ts#L12-L25
`Message`, and the wire format of the message, does not store information that allows `AccountMeta` to be recreated per instruction. Only the number of signatures required, and the numbers of read-only accounts (signed and unsigned) are stored in the message header.

2. https://github.com/solana-labs/solana/blob/17cc095d2892654225b5efe91b55c19eb932bfd3/web3.js/src/transaction.ts#L705-L709
Because of 1), when a Transaction is populated from a `Message`, if an account has `isSigner` or `isWritable` flags for any instruction, these flags will be propagated to the account for every instruction in the transaction.  

3. https://github.com/solana-labs/solana/blob/17cc095d2892654225b5efe91b55c19eb932bfd3/web3.js/src/transaction.ts#L261-L270
Because of 2), when `AccountMeta` are sorted by `isSigner` and `isWritable`, accounts that didn’t have `isSigner` or `isWritable` flags before serialization may have them after deserialization. The order of these is likely to change. 

4. https://github.com/solana-labs/solana/blob/17cc095d2892654225b5efe91b55c19eb932bfd3/web3.js/src/transaction.ts#L272-L285
Because of 3), `AccountMeta` will be iterated over in a different order to gather the unique keys. 

5. https://github.com/solana-labs/solana/blob/17cc095d2892654225b5efe91b55c19eb932bfd3/web3.js/src/transaction.ts#L327-L343
Because of 4), the signed and unsigned keys will be collected in a different order and concatenated.

6. https://github.com/solana-labs/solana/blob/17cc095d2892654225b5efe91b55c19eb932bfd3/web3.js/src/message.ts#L191-L196
Because of 5), the keys will be serialized as bytes in a different order. When these bytes are signed, this will result in a different signature.

In short, because of 1), `Transaction` to `Message` performs lossy compression. You can’t deterministically serialize a `Transaction` to a `Message` and deserialize this `Message` back to the original `Transaction`.

**Comment by jstarry** [comment]
> I dug into the code and came up with what I think is causing t

Cool, this is what I was referring above when I said that "The way Transaction.populate is decompiling the message account keys is losing the ordering used by the processed transaction so I think we have to get rid of that."

So I think we're all in agreement that `Transaction.populate` needs to go away?

**Comment by t-nelson** [comment]
I think most of the `Transaction` API needs to go away.  It's nearly all a broken wrapper for things that should be on `Message`.  Such a refactor is alluded to [here](https://github.com/solana-labs/solana/issues/23491).  Seems like we might additionally need a second `Message` type that can be deserialized to, but not modified in any way? 

So we'd have something like...
```
Message --> CompiledMessage <+-> bytes
                             +-> Transaction
```

**Comment by vpontis** [comment]
> Cool, this is what I was referring above when I said that "The way Transaction.populate is decompiling the message account keys is losing the ordering used by the processed transaction so I think we have to get rid of that."

Thanks for investigating this all! @jstarry fwiw, I missed the implications of your comment so it was helpful to have @jordansexton's longer explanation. 

> So I think we're all in agreement that `Transaction.populate` needs to go away?

I wonder if it makes sense to remove the `class Message` abstraction and instead merge the message into the transaction. 

Btw, another thing that @jordansexton and I were talking about is that when you set an account to not writable for an instruction, you expect it to stay that way. But if another instruction increases permission levels (makes an account writable) then you lose that guarantee in your first instruction.

**Comment by t-nelson** [comment]
> I wonder if it makes sense to remove the class Message abstraction and instead merge the message into the transaction.

No. Wrong direction.  `Transaction` is a container for a `Message` and its signatures, nothing more

**Comment by jordansexton** [comment]
> So I think we're all in agreement that `Transaction.populate` needs to go away?

What if we throw an error if a `Transaction` that was created with `populate` is compiled for signing again, or don't recompile the message to sign it?

I think populating a transaction from a message is still valuable for inspecting the instructions, fee payer, etc. but it needs to be clear that you can't sign it that way.

**Comment by jordansexton** [comment]
PR attempt @ #23720, please review!

**Comment by ftelnov** [comment]
So what's about this issue? As for me, I'm still getting "Signature verification failed" sometimes. It's very strange, sometimes transactions are passed, sometimes not, despite the fact that algorithm of building on backend-side is the same in all the cases. 
I'm as well using offline signing, forming and serialising transaction on back-side, deserialising and confirming on frontend-side. And this lead to errors. 
I saw the workaround above, but that didn't help in my case at all. On backend I added Transaction.from(tx.serialize(...)), but on frontend - still signature errors sometimes.
Pls, tell me, maybe someone solved it through provided workaround? @apeTim?

**Comment by ftelnov** [comment]
> I think populating a transaction from a message is still valuable for inspecting the instructions, fee payer, etc. but it needs to be clear that you can't sign it that way.
> 
> >
Do you mean that signing "pre-signed -> serialized -> deserilized" transaction is the dead-concept at all? Maybe you can suggest something else? Sorry, I just trying to overcome this bug, but can't find any 100% working solution.

**Comment by jordansexton** [comment]
Fix should be released in `@solana/web3.js@1.39.1`

**Comment by juchiast** [comment]
I'm still seeing a similar issue when signing with backpack and phantom wallets

https://github.com/orgs/phantom/discussions/162
